File: iOS/TUILiveKit/Source/LiveRoom/Core/EngineObserver.swift
//
//  EngineObserver.swift
//  TUILiveKit
//
//  Created by krabyu on 2023/11/6.
//

import RTCRoomEngine
import UIKit
#if canImport(TXLiteAVSDK_TRTC)
    import TXLiteAVSDK_TRTC
#elseif canImport(TXLiteAVSDK_Professional)
    import TXLiteAVSDK_Professional
#endif

class EngineObserver: NSObject {
    private var liveRoomInfo: LiveRoomInfo
    @WeakLazyInjected var engineManager:EngineServiceProvider?
    private var engineService: RoomEngineService? {
        return engineManager?.getRoomEngineService(roomId: liveRoomInfo.roomId.value)
    }
    
    init(liveRoomInfo:LiveRoomInfo) {
        self.liveRoomInfo = liveRoomInfo
    }

    private func handleLeftList(leftList: [TUISeatInfo]) {
        guard let engineService = engineService else{ return}
        for seatInfo: TUISeatInfo in leftList {
            guard let userId = seatInfo.userId else { continue }
            if userId == engineService.liveRoomInfo.selfInfo.userId {
                engineService.liveRoomInfo.selfInfo.role.value = .audience
                engineService.liveRoomInfo.selfInfo.status.value = .none
                return
            }
        }
    }
}

extension EngineObserver: TUIRoomObserver {
    func onRemoteUserEnterRoom(roomId: String, userInfo: TUIUserInfo) {
        LiveKitLog.info("\(#file)","\(#line)","onRemoteUserEnterRoom:[roomId:\(roomId),userId:\(userInfo.userId)]")
        let user = liveRoomInfo.userInfoPool.get(userInfo.userId)
        user.update(userInfo)
        guard liveRoomInfo.audienceList.value.filter({ $0.userId == userInfo.userId }).first != nil else {
            liveRoomInfo.audienceList.value.append(user)
            return
        }
    }

    func onRemoteUserLeaveRoom(roomId: String, userInfo: TUIUserInfo) {
        guard let engineService = engineService else{ return}
        liveRoomInfo.audienceList.value = liveRoomInfo.audienceList.value.filter({ $0.userId != userInfo.userId })
        liveRoomInfo.linkingAudienceList.value = liveRoomInfo.linkingAudienceList.value.filter({ $0.userId != userInfo.userId })
        engineService.liveRoomInfo.applyLinkAudienceList =
        engineService.liveRoomInfo.applyLinkAudienceList.filter({ $0.userId != userInfo.userId })
        liveRoomInfo.userInfoPool.remove(userInfo.userId)
        LiveKitLog.info("\(#file)","\(#line)","onRemoteUserLeaveRoom:[roomId:\(roomId),userId:\(userInfo.userId)]")
    }

    func onUserAudioStateChanged(userId: String, hasAudio: Bool, reason: TUIChangeReason) {
        LiveKitLog.info("\(#file)","\(#line)","onUserAudioStateChanged:[userId:\(userId),hasAudio:\(hasAudio)]")
        let user = liveRoomInfo.userInfoPool.get(userId)
        user.audioInfo.muteAudio.value = !hasAudio
    }

    func onUserVideoStateChanged(userId: String, streamType: TUIVideoStreamType, hasVideo: Bool, reason: TUIChangeReason) {
        LiveKitLog.info("\(#file)","\(#line)","onUserVideoStateChanged:[userId:\(userId),hasVideo:\(hasVideo)]")
        let user = liveRoomInfo.userInfoPool.get(userId)
        user.videoInfo.isCameraOpened.value = hasVideo
    }

    func onSeatListChanged(seatList: [TUISeatInfo], seated seatedList: [TUISeatInfo], left leftList: [TUISeatInfo]) {
        LiveKitLog.info("\(#file)","\(#line)","onSeatListChanged:[seatList.count:\(seatList.count),leftList.count:\(leftList.count)]")
        engineService?.handleSeatList(seatList: seatList,seatedList: seatedList)
        handleLeftList(leftList: leftList)
    }

    func onRoomDismissed(roomId: String) {
        LiveKitLog.info("\(#file)","\(#line)","onRoomDismissed:[roomId:\(roomId)]")
        engineService?.changeUserLiveStatus(userLiveStatus: .none)
    }

    // MARK: - Signaling request-related callback

    func onRequestReceived(request: TUIRequest) {
        LiveKitLog.info("\(#file)","\(#line)","onRequestReceived:[requestId:\(request.requestId),requestAction:\(request.requestAction)]")
        guard let userInfo = liveRoomInfo.audienceList.value.first(where: { $0.userId == request.userId }) else { return }
        guard let engineService = engineService else{ return }
        userInfo.requestId = request.requestId
        if request.requestAction == .takeSeat && liveRoomInfo.interactionType.value == .link {
            var array = engineService.liveRoomInfo.applyLinkAudienceList.filter({ $0.userId != userInfo.userId })
            array.append(userInfo)
            engineService.liveRoomInfo.applyLinkAudienceList = array
        }
    }

    func onRequestCancelled(requestId: String, userId: String) {
        LiveKitLog.info("\(#file)","\(#line)","onRequestCancelled:[requestId:\(requestId),userId:\(userId)]")
        guard let engineService = engineService else{ return }
        engineService.liveRoomInfo.applyLinkAudienceList =
        engineService.liveRoomInfo.applyLinkAudienceList.filter({ $0.requestId != requestId })
    }
    
    func onRoomUserCountChanged(roomId: String, userCount: Int) {
        guard let engineService = engineService else{ return }
        let count = max(userCount - 1, 0)
        engineService.liveRoomInfo.audienceCount.value = max(count, engineService.liveRoomInfo.audienceCount.value)
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/liveroom/core/EngineObserver.java
package com.trtc.uikit.livekit.liveroom.core;

import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.tencent.cloud.tuikit.engine.room.TUIRoomObserver;
import com.trtc.uikit.livekit.common.core.store.LiveStore;
import com.trtc.uikit.livekit.common.utils.LiveKitLog;
import com.trtc.uikit.livekit.liveroom.data.LiveRoomInfo;
import com.trtc.uikit.livekit.liveroom.data.UserInfo;

import java.util.List;

public class EngineObserver extends TUIRoomObserver {

    private final LiveRoomInfo      mLiveRoomInfo;
    private final RoomEngineService mRoomEngineService;

    public EngineObserver(LiveRoomInfo roomInfo, RoomEngineService service) {
        mLiveRoomInfo = roomInfo;
        mRoomEngineService = service;
    }

    @Override
    public void onRemoteUserEnterRoom(String roomId, TUIRoomDefine.UserInfo userInfo) {
        LiveKitLog.info("EngineObserver onRemoteUserEnterRoom:[roomId:" + roomId + ",userInfo:" + userInfo + "]");
        UserInfo user = null;
        for (UserInfo info : mLiveRoomInfo.audienceList.get()) {
            if (info.userId.equals(userInfo.userId)) {
                user = info;
                break;
            }
        }
        if (user == null) {
            mLiveRoomInfo.audienceList.add(new UserInfo(userInfo));
        } else {
            user.update(userInfo);
        }
    }

    @Override
    public void onRemoteUserLeaveRoom(String roomId, TUIRoomDefine.UserInfo userInfo) {
        LiveKitLog.info("EngineObserver onRemoteUserLeaveRoom:[roomId:" + roomId + ",userInfo:" + userInfo + "]");
        UserInfo info = new UserInfo(userInfo);
        mLiveRoomInfo.audienceList.remove(info);
        LiveKitStore.sharedInstance().applyLinkAudienceList.remove(info);
    }

    @Override
    public void onRoomDismissed(String roomId) {
        LiveKitLog.info("EngineObserver onRoomDismissed:[roomId:" + roomId + "]");
        mLiveRoomInfo.userLiveStatus.set(TUILiveDefine.UserLiveStatus.DASHBOARD);
    }

    @Override
    public void onUserAudioStateChanged(String userId, boolean hasAudio, TUIRoomDefine.ChangeReason reason) {
        LiveKitLog.info("EngineObserver onUserAudioStateChanged:[userId:" + userId + ",hasAudio:" + hasAudio
                + ",reason:" + reason + "]");
        if (userId.equals(mLiveRoomInfo.anchorInfo.userId)) {
            mLiveRoomInfo.anchorInfo.audioInfo.muteAudio.set(!hasAudio);
        }

        for (UserInfo user : mLiveRoomInfo.linkingAudienceList.get()) {
            if (user.userId.equals(userId)) {
                user.audioInfo.muteAudio.set(!hasAudio);
                break;
            }
        }

        if (hasAudio) {
            mLiveRoomInfo.hasAudioList.add(userId);
        } else {
            mLiveRoomInfo.hasAudioList.remove(userId);
        }
    }

    @Override
    public void onUserVideoStateChanged(String userId, TUIRoomDefine.VideoStreamType streamType, boolean hasVideo,
                                        TUIRoomDefine.ChangeReason reason) {
        LiveKitLog.info("EngineObserver onUserVideoStateChanged:[userId:" + userId + ",streamType:" + streamType
                + ",hasVideo:" + hasVideo + ",reason:" + reason + "]");
        if (userId.equals(mLiveRoomInfo.anchorInfo.userId)) {
            mLiveRoomInfo.anchorInfo.videoInfo.isCameraOpened.set(hasVideo);
        }

        for (UserInfo user : mLiveRoomInfo.linkingAudienceList.get()) {
            if (user.userId.equals(userId)) {
                user.videoInfo.isCameraOpened.set(hasVideo);
                break;
            }
        }

        if (hasVideo) {
            mLiveRoomInfo.hasVideoList.add(userId);
        } else {
            mLiveRoomInfo.hasVideoList.remove(userId);
        }
    }

    @Override
    public void onSeatListChanged(List<TUIRoomDefine.SeatInfo> seatList, List<TUIRoomDefine.SeatInfo> seatedList,
                                  List<TUIRoomDefine.SeatInfo> leftList) {
        LiveKitLog.info("EngineObserver onSeatListChanged:[seatList=" + seatList + ",seatedList:" + seatedList
                + ",leftList:" + leftList + "]");
        mRoomEngineService.handleSeatChange(seatList, seatedList);
    }

    @Override
    public void onRequestReceived(TUIRoomDefine.Request request) {
        LiveKitLog.info("EngineObserver onRequestReceived:[request=" + request + "]");
        if (request.requestAction == TUIRoomDefine.RequestAction.REQUEST_TO_TAKE_SEAT) {
            UserInfo userInfo = new UserInfo(request.userId);
            userInfo.requestId = request.requestId;
            userInfo.name.set(request.userName);
            userInfo.avatarUrl.set(request.avatarUrl);
            LiveKitStore.sharedInstance().applyLinkAudienceList.add(userInfo);
        }
    }

    @Override
    public void onRequestCancelled(String requestId, String userId) {
        LiveKitLog.info("EngineObserver onRequestCancelled:[requestId=" + requestId + ",userId:" + userId + "]");
        for (UserInfo info : LiveKitStore.sharedInstance().applyLinkAudienceList.get()) {
            if (info.userId.equals(userId)) {
                LiveKitStore.sharedInstance().applyLinkAudienceList.remove(info);
                break;
            }
        }
    }

    @Override
    public void onRoomUserCountChanged(String roomId, int userCount) {
        LiveKitLog.info("EngineObserver onRoomUserCountChanged:[roomId=" + roomId + ",userCount:" + userCount + "]");
        // Anchor are not counted in the number of viewers
        userCount--;
        if (userCount > 0) {
            mLiveRoomInfo.audienceCount.set(userCount);
            if (userCount > LiveStore.sharedInstance().getLiveController().getRoomSate().maxAudienceNumber) {
                LiveStore.sharedInstance().getLiveController().getRoomController().updateMaxAudienceNumber(userCount);
            }
        }
    }
}
